components/authz-service/server/v2/policy_refresher.go

```diff
@@ -0,0 +1,276 @@
+)
+
+var ErrMessageBoxFull = errors.New("Message box full")
+
+type PolicyRefresher interface {
+	Refresh(ctx context.Context) error
+	RefreshAsync() error
+}
+
+type policyRefersher struct {
+	log                      logger.Logger
+	store                    storage.Storage
+	engine                   engine.V2Writer
+	refreshRequests          chan policyRefresherMessageRefresh
+	antiEntropyTimerDuration time.Duration
+}
+
+type policyRefresherMessageRefresh struct {
+	ctx    context.Context
+	status chan error
+}
+
+func (m *policyRefresherMessageRefresh) Respond(err error) {
+	if m.status != nil {
+		select {
+		case m.status <- err:
+		default:
+		}
+		close(m.status)
+	}
+}
+
+func (m *policyRefresherMessageRefresh) Err() error {
+	return <-m.status
+}
+
+func NewPolicyRefresher(ctx context.Context, log logger.Logger, store storage.Storage,
+	engine engine.V2Writer) PolicyRefresher {
+	refresher := &policyRefersher{
+		log:                      log,
+		store:                    store,
+		engine:                   engine,
+		refreshRequests:          make(chan policyRefresherMessageRefresh, 1),
+		antiEntropyTimerDuration: 10 * time.Second,
+	}
+	go refresher.run(ctx)
+	return refresher
+}
+
+func (refresher *policyRefersher) run(ctx context.Context) {
+	lastPolicyID := ""
+	antiEntropyTimer := time.NewTimer(refresher.antiEntropyTimerDuration)
+RUNLOOP:
+	for {
+		select {
+		case <-ctx.Done():
+			refresher.log.WithError(ctx.Err()).Info("Policy refresher exiting")
+			break RUNLOOP
+		case m := <-refresher.refreshRequests:
+			var err error
+			lastPolicyID, err = refresher.refresh(m.ctx, lastPolicyID)
+			m.Respond(err)
+			if !antiEntropyTimer.Stop() {
+				<-antiEntropyTimer.C
+			}
+		case <-antiEntropyTimer.C:
+			var err error
+			lastPolicyID, err = refresher.refresh(ctx, lastPolicyID)
+			if err != nil {
+				refresher.log.WithError(err).Warn("Anti-entropy refresh failed")
+			}
+		}
+
+		antiEntropyTimer.Reset(refresher.antiEntropyTimerDuration)
+	}
+	close(refresher.refreshRequests)
+}
+
+func (refresher *policyRefersher) refresh(ctx context.Context, lastPolicyID string) (string, error) {
+	curPolicyID, err := refresher.store.GetPolicyChangeID(ctx)
+	if err != nil {
+		refresher.log.WithError(err).Warn("Failed to get current policy change ID")
+		return lastPolicyID, err
+	}
+	if curPolicyID != lastPolicyID {
+		refresher.log.WithFields(logrus.Fields{
+			"lastPolicyID": lastPolicyID,
+			"curPolicyID":  curPolicyID,
+		}).Info("Refreshing engine store")
+
+		if err := refresher.updateEngineStore(ctx); err != nil {
+			refresher.log.WithError(err).Warn("Failed to refresh engine store")
+			return lastPolicyID, err
+		}
+	}
+	return curPolicyID, nil
+}
+
+func (refresher *policyRefersher) Refresh(ctx context.Context) error {
+	m := policyRefresherMessageRefresh{
+		ctx:    ctx,
+		status: make(chan error, 1),
+	}
+	refresher.refreshRequests <- m
+	return m.Err()
+}
+
+func (refresher *policyRefersher) RefreshAsync() error {
```

Comment:
Is this used?

Reply:
nope. I was going to use it for the pg notify stuff, but i'm not going to do that in this pr